sphinx/domains/python.py

```diff
@@ -285,12 +290,54 @@ def add_target_and_index(self, name_cls, sig, signode):
                                               fullname, '', None))
 
     def before_content(self):
-        # needed for automatic qualification of members (reset in subclasses)
-        self.clsname_set = False
+        # type: () -> None
+        """Handle object nesting before content
+
+        If this class is a nestable object, such as a class object, build up a
+        representation of the nesting heirarchy so that de-nesting multiple
+        levels works correctly.
+
+        If this class isn't a nestable object, just set the current class name
+        using the object prefix, if any. This class name will be removed with
+        :py:meth:`after_content`, and is not added to the list of nested
+        classes.
+        """
+        prefix = None
+        if self.names:
+            (cls_name, cls_name_prefix) = self.names.pop()
```

Comment:
Hmm... `self.names.pop()` contains always "class name" ?
As a quick look, it also contains modules, methods, functions and so on.
If my consideration is true, please rename this to better one.